rdmd.d

```diff
@@ -916,3 +891,71 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function or affected by dryRun. Append the string "IfLive" to the end
+of the function for it to be skipped during a dry run.
+
+These functions allow the filename to be given once so the log statement
+always matches the operation. Using it also guarantees you won't forget to
+include a `yap` alongside any file operation you want to be logged.
+*/
+struct Filesystem
+{
+static:
+    static auto opDispatch(string func, T...)(T args)
+    {
+        static if (func.endsWith("IfLive"))
+        {
+            enum fileFunc = func[0 .. $ - "IfLive".length];
+            enum skipOnDryRun = true;
+        }
+        else
+        {
+            enum fileFunc = func;
+            enum skipOnDryRun = false;
+        }
+
+        static if ([fileFunc].findAmong(["exists", "timeLastModified", "isFile", "existsAsFile"]))
+            yap("stat ", args[0]);
+        else static if (fileFunc == "rename")
+            yap("mv ", args[0], " ", args[1]);
+        else static if ([fileFunc].findAmong(["remove", "mkdirRecurse", "rmdirRecurse", "dirEntries", "write", "touch"]))
+            yap(fileFunc, " ", args[0]);
+        else static assert(0, "Filesystem.opDispath has not implemented " ~ fileFunc);
+
+        static if (skipOnDryRun)
+        {
+            if (dryRun)
+                return;
+        }
```

Comment:
This is a leaky abstraction: it assumes that the functions which need the dry-run option have `void` return-type.  This may be true of the particular functions currently used, but is not guaranteed to be generally true.

Reply:
Yes I thought about this.  I original was going to return the `.init` value of the return type, but then I realized that was assuming the the `.init` value would have been representative of a return value when NOT EXECUTING the function.  After realizing this, it became clear that any function affected by dry run would most certainly have a `void` return value since a function that can optionally do nothing would return nothing.

Reply:
We'll get a compilation error in that case, so I think we're good here.

Reply:
> We'll get a compilation error in that case

Well, of course.  But it'd be nice to avoid the maintenance burden of some future developer wanting to add an extra function here, and finding that it breaks the assumptions made by `Filesystem.opDispatch`.

Reply:
I think we are good here too. It's a single file, has produced < 5 bugs in two years and is not a huge codebase anyways. Also doing a  `__traits(allMembers` over `std.file` would lead to worse pretty-printed loggin (I thought about proposing it too).

Reply:
You've not addressed my reasoning for assuming the function must return `void`.